### ISSUE TITLE
index_copy_ still costs O(destination) for a vLLM-shaped destination (follow-up to #4239)

### ISSUE THREAD

**Issue by gkumbhat** [torch-spyre/torch-spyre#4608]
### What happened?

#4239 ("index_copy_ costs O(destination) instead of O(payload)") was closed as
fixed by #4163, with the fix reported as applying to *layout-compliant*
destinations. A destination shaped the way vLLM allocates attention output still
scales with the destination rather than the payload.

The destination here is a `[tokens, heads, head_size]` **view of a 2-D
`[tokens, heads*head_size]` allocation** -- which is exactly how
`Attention.forward` builds it: `torch.empty(tokens, heads*head_size).view(-1,
heads, head_size)`.

Ratios (`repro_index_copy_destination_cost.py`):

| direction | change | relative cost |
|---|---|---|
| A: payload fixed at 512 rows, destination grows | destination **8x** larger | **~2.26x** |
| B: destination fixed at 4096 rows, payload grows | payload **64x** larger | **~1.42x** |

An 8x larger destination costs noticeably more than a 64x larger payload, so the
scatter is still being charged for the whole destination.

### What did you expect to happen?

Either extend the fast path to cover this destination shape, or document what
makes a destination compliant so callers can satisfy it deliberately.

From reading `_inductor/enforce_indirect_access_layout.py`, the requirement
appears to be that the scattered dim sits outermost -- and ours *is* dim 0, so it
is not obvious why we do not qualify. `hbm_pool_planning.py` mentions
`_insert_mutation_relayout_copy` inserting a copy for a non-compliant scatter,
which would explain the scaling exactly. We tried to confirm via the
`scatter_destination_check` debug line but could not get that logging to surface.

### How can we reproduce it?

[repro_index_copy_destination_cost.py](https://github.com/user-attachments/files/32310074/repro_index_copy_destination_cost.py)

### Any environmental details we need to know?

Used latest main of torch-spyre (sept 16)

### Related

- #4239, #4163 -- the original issue and its fix
- Companion issue: `index_select` cost is dominated by the source size: #4607 

### Relevant log output

```Shell

```

**Comment by avery-blanchard** [comment]
@gkumbhat - the layout here does not fit the constraints required for indirect access from the backend. A copy is inserted by the compile to enforce these layout constraints before doing the index copy. Here is an updated version of the reproducer that has a compliant layout.

```
import time

import torch
import torch_spyre  # noqa: F401  (registers the 'spyre' device)
from torch_spyre._C import SpyreTensorLayout, get_device_dtype, get_elem_in_stick

DEV, DT = "spyre", torch.float16
HEADS, HEAD_SIZE, ITERS = 12, 64, 30


def scatter(dst, idx, payload):
    dst.index_copy_(0, idx, payload)
    return dst


def slot_major_layout(rows: int) -> SpyreTensorLayout:
    """Device layout with the scattered dim (rows) at device position 0."""
    eps = get_elem_in_stick(DT)
    sticks = HEAD_SIZE // eps
    return SpyreTensorLayout([rows, HEADS, sticks, eps],
        [HEADS * sticks * eps, sticks * eps, eps, 1],
        get_device_dtype(DT))

def vllm_style_dest(rows):
    """Allocated directly in the (rows, HEADS, HEAD_SIZE) shape, with an
    explicit slot-major device layout, so the indirect-access dim-order
    constraint is already satisfied and no relayout copy is inserted."""
    dst = torch.zeros((rows, HEADS, HEAD_SIZE), dtype=DT)
    return dst.to(DEV, device_layout=slot_major_layout(rows))


def bench(fn, iters=ITERS):
    fn()
    t0 = time.perf_counter()
    for _ in range(iters):
        fn()
    return (time.perf_counter() - t0) / iters


def main() -> None:
    s = torch.compile(scatter, dynamic=False)
    with torch.inference_mode():
        print("A. payload fixed at 512 rows, destination grows")
        print(f"{'dest rows':>10} {'relative cost':>14}")
        base = None
        for rows in (512, 1024, 2048, 4096):
            idx = torch.arange(512, dtype=torch.int32).to(DEV)
            payload = torch.randn(512, HEADS, HEAD_SIZE, dtype=DT).to(DEV)
            dst = vllm_style_dest(rows)

            s(dst, idx, payload)
            t = bench(lambda: s(dst, idx, payload))
            base = base or t
            print(f"{rows:>10} {t / base:>13.2f}x")

        print("\nB. destination fixed at 4096 rows, payload grows")
        print(f"{'payload rows':>12} {'relative cost':>14}")
        dst = vllm_style_dest(4096)
        base = None
        for rows in (64, 512, 2048, 4096):
            idx = torch.arange(rows, dtype=torch.int32).to(DEV)
            payload = torch.randn(rows, HEADS, HEAD_SIZE, dtype=DT).to(DEV)
            s(dst, idx, payload)
            t = bench(lambda: s(dst, idx, payload))
            base = base or t
            print(f"{rows:>12} {t / base:>13.2f}x")
    print("done")


if __name__ == "__main__":
    main()

```
